feature_embeddings.py

```diff
@@ -0,0 +1,149 @@
+
+    def create_tag_dict(self):
+        # convert the tags to indices and create dict for conversion
+        tag_values = list(set(self.data["tag"].values))
+        tag_values.append("PAD")
+        tag2idx = {t: i for i, t in enumerate(tag_values)}
+        return tag2idx
+
+    def tokenize_and_preserve_labels(self, sentence, text_labels):
+        """
+        tokenizes the words into the bert word elements and duplicates the labels
+        :return:
+        """
+        tokenized_sentence = []
+        labels = []
+
+        for word, label in zip(sentence, text_labels):
+
+            # Tokenize the word and count # of subwords the word is broken into
+            tokenized_word = self.tokenizer.tokenize(word)
+            n_subwords = len(tokenized_word)
+
+            # Add the tokenized word to the final tokenized word list
+            tokenized_sentence.extend(tokenized_word)
+
+            # Add the same label to the new list of labels `n_subwords` times
+            labels.extend([label] * n_subwords)
+
+        return tokenized_sentence, labels
+
+
+    def prep_bert_inputs(self):
+        # get the sentences and the labels from the dataframe
+        getter = SentenceGetter(self.data)
+        sentences = [[s[0] for s in sent] for sent in getter.sentences]
+        labels = [[s[1] for s in sent] for sent in getter.sentences]
+
+        # tokenize the words + duplicate the labels --> list of tuples with (sent tokens, labels)
+        tokenized_texts_and_labels = [
+            self.tokenize_and_preserve_labels(sent, labs)
+            for sent, labs in zip(sentences, labels)
+        ]
+
+        # create seperate vars for the tokenized text and their labels
+        tokenized_texts = [token_label_pair[0] for token_label_pair in tokenized_texts_and_labels]
+        labels = [token_label_pair[1] for token_label_pair in tokenized_texts_and_labels]
+
+        # turn vars to numericals
+        all_ids = [self.tokenizer.convert_tokens_to_ids(txt) for txt in tokenized_texts]
+        all_target_tag = [[self.tag2idx.get(l) for l in lab] for lab in labels]
+
+        # cut/pad the sequences & the bert markers
+        for i in range(len(all_ids)):
+            # todo: hmm this is just cutting off everything after max length
+
+            # cut until max length, minus 2 to make room for bert markers
+            ids = all_ids[i][:self.max_len - 2]
+            target_tag = all_target_tag[i][:self.max_len - 2]
+
+            # bert markers and paddings for the rest
+            ids = [101] + ids + [102]
+            target_tag = [self.tag2idx["PAD"]] + target_tag + [self.tag2idx["PAD"]]
+
+            # pad shortened sequences
+            padding_len = self.max_len - len(ids)
+            all_ids[i] = ids + ([0] * padding_len)
+            all_target_tag[i] = target_tag + ([self.tag2idx["PAD"]] * padding_len)
+
+        # creating the attention mask
+        attention_mask = [[float(i != 0.0) for i in ii] for ii in all_ids]

```

Comment:
This is a bit dangerous, you should never compare floats like this, but here is float(int != float)?? AA bit sketchy line which may bring unexpected output